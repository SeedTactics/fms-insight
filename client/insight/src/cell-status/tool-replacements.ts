/* Copyright (c) 2022, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { addDays } from "date-fns";
import { ILogEntry, LogType, ToolUse } from "../network/api.js";
import { LazySeq, HashMap, OrderedMap, HashableObj, hashValues } from "@seedtactics/immutable-collections";
import { durationToMinutes } from "../util/parseISODuration.js";
import type { ServerEventAndTime } from "./loading.js";
import { atom } from "jotai";

export type ToolReplacement =
  | {
      readonly tool: string;
      readonly pocket: number;
      readonly type: "ReplaceBeforeCycleStart";
      readonly useAtReplacement: number | null;
      readonly cntAtReplacement: number | null;
    }
  | {
      readonly tool: string;
      readonly pocket: number;
      readonly type: "ReplaceInCycle";

      // when a replacement happens in the middle of the cycle, we don't know exactly how much usage there was.
      // We instead estimate it by recording the total use at the beginning and end of the cycle and estimate using
      // the calculated average usage (from tool-usage.ts)

      // Thus, useAtReplacement estimate = totalUseAtBeginningOfCycle + (averageUse - totalUseAtEndOfCycle)
      readonly totalUseAtBeginningOfCycle: number | null;
      readonly totalUseAtEndOfCycle: number | null;
      readonly totalCntAtBeginningOfCycle: number | null;
      readonly totalCntAtEndOfCycle: number | null;
    };

export type ToolReplacements = {
  readonly time: Date;
  readonly replacements: ReadonlyArray<ToolReplacement>;
};

export class StationGroupAndNum implements HashableObj {
  readonly group: string;
  readonly num: number;
  constructor(group: string, num: number) {
    this.group = group;
    this.num = num;
  }
  static ofLogCycle(e: Readonly<ILogEntry>): StationGroupAndNum {
    return new StationGroupAndNum(e.loc, e.locnum);
  }
  hash(): number {
    return hashValues(this.group, this.num);
  }
  compare(other: StationGroupAndNum): number {
    return this.group === other.group ? this.num - other.num : this.group.localeCompare(other.group);
  }
  toString(): string {
    return this.group + " #" + this.num.toString();
  }
}

export type ToolReplacementsByCntr = OrderedMap<number, ToolReplacements>;
export type ToolReplacementsByStation = OrderedMap<StationGroupAndNum, ToolReplacementsByCntr>;
type MostRecentUseByStation = HashMap<StationGroupAndNum, ReadonlyArray<ToolUse>>;

type ReplacementsAndMostRecentUse = {
  readonly replacements: ToolReplacementsByStation;
  readonly recentUse: MostRecentUseByStation;
};

const emptyReplacementsAndUse: ReplacementsAndMostRecentUse = {
  replacements: OrderedMap.empty(),
  recentUse: HashMap.empty(),
};

const last30ToolReplacementsRW = atom<ReplacementsAndMostRecentUse>(emptyReplacementsAndUse);
export const last30ToolReplacements = atom<ToolReplacementsByStation>(
  (get) => get(last30ToolReplacementsRW).replacements
);

const specificMonthToolReplacementsRW = atom<ReplacementsAndMostRecentUse>(emptyReplacementsAndUse);
export const specificMonthToolReplacements = atom<ToolReplacementsByStation>(
  (get) => get(specificMonthToolReplacementsRW).replacements
);

function addReplacementsFromLog(
  old: ReplacementsAndMostRecentUse,
  e: Readonly<ILogEntry>
): ReplacementsAndMostRecentUse {
  const key = StationGroupAndNum.ofLogCycle(e);
  const newReplacements = old.replacements.alter(key, (cycles) => {
    const replacements: Array<ToolReplacement> = [];
    for (const use of e.tooluse ?? []) {
      // see server/lib/BlackMaple.MachineFramework/backend/ToolSnapshotDiff.cs for the original calculations
      const useDuring =
        use.toolUseDuringCycle !== null && use.toolUseDuringCycle !== undefined
          ? durationToMinutes(use.toolUseDuringCycle)
          : null;
      const totalUseAtEnd =
        use.totalToolUseAtEndOfCycle !== null && use.totalToolUseAtEndOfCycle !== undefined
          ? durationToMinutes(use.totalToolUseAtEndOfCycle)
          : null;
      const cntDuring = use.toolUseCountDuringCycle ?? null;
      const totalCntAtEnd = use.totalToolUseCountAtEndOfCycle ?? null;

      if ((useDuring || cntDuring) && useDuring === totalUseAtEnd && cntDuring === totalCntAtEnd) {
        // replace before cycle start
        const last = old.recentUse.get(key)?.find((e) => e.tool === use.tool && e.pocket === use.pocket);
        if (last) {
          const lastTotalUse = last.totalToolUseAtEndOfCycle
            ? durationToMinutes(last.totalToolUseAtEndOfCycle)
            : null;
          if (lastTotalUse || last.totalToolUseCountAtEndOfCycle) {
            replacements.push({
              tool: use.tool,
              pocket: use.pocket,
              type: "ReplaceBeforeCycleStart",
              useAtReplacement: lastTotalUse,
              cntAtReplacement: last.totalToolUseCountAtEndOfCycle ?? null,
            });
          }
        }
      } else if (
        use.toolChangeOccurred &&
        ((useDuring && use.configuredToolLife) || (cntDuring && use.configuredToolLifeCount))
      ) {
        // replace in the middle of the cycle

        // the server calculates  useDuring = lifetime - useAtStartOfCycle + useAtEndOfCycle
        // so solving for useAtStart = lifetime - useDuring + useAtEndOfCycle
        const lifetime =
          use.configuredToolLife !== null && use.configuredToolLife !== undefined
            ? durationToMinutes(use.configuredToolLife)
            : null;
        const lifeCnt = use.configuredToolLifeCount ?? null;
        replacements.push({
          tool: use.tool,
          pocket: use.pocket,
          type: "ReplaceInCycle",
          totalUseAtBeginningOfCycle:
            lifetime !== null && useDuring !== null ? lifetime - useDuring + (totalUseAtEnd ?? 0) : null,
          totalUseAtEndOfCycle: totalUseAtEnd,
          totalCntAtBeginningOfCycle:
            lifeCnt !== null && cntDuring !== null ? lifeCnt - cntDuring + (totalCntAtEnd ?? 0) : null,
          totalCntAtEndOfCycle: totalCntAtEnd,
        });
      }
    }
    if (replacements.length > 0) {
      return (cycles ?? OrderedMap.empty()).set(e.counter, { time: e.endUTC, replacements });
    } else {
      return cycles;
    }
  });

  const newRecent = e.tooluse ? old.recentUse.set(key, e.tooluse) : old.recentUse;

  return { replacements: newReplacements, recentUse: newRecent };
}

export const setLast30ToolReplacements = atom(null, (_, set, log: ReadonlyArray<Readonly<ILogEntry>>) => {
  set(last30ToolReplacementsRW, (oldCycles) =>
    LazySeq.of(log)
      .filter(
        (e) => e.type === LogType.MachineCycle && !e.startofcycle && !!e.tooluse && e.tooluse.length > 0
      )
      .foldLeft(oldCycles, addReplacementsFromLog)
  );
});

export const updateLastToolReplacements = atom(null, (_, set, { evt, now, expire }: ServerEventAndTime) => {
  if (
    evt.logEntry &&
    !evt.logEntry.startofcycle &&
    evt.logEntry.type === LogType.MachineCycle &&
    evt.logEntry.tooluse &&
    evt.logEntry.tooluse.length > 0
  ) {
    const log = evt.logEntry;

    set(last30ToolReplacementsRW, (oldCycles) => {
      if (expire) {
        const thirtyDaysAgo = addDays(now, -30);
        const newReplace = oldCycles.replacements.collectValues((es) => {
          const newEs = es.filter((e) => e.time >= thirtyDaysAgo);
          return newEs.size > 0 ? newEs : null;
        });
        oldCycles = { replacements: newReplace, recentUse: oldCycles.recentUse };
      }

      return addReplacementsFromLog(oldCycles, log);
    });
  }
});

export const setSpecificMonthToolReplacements = atom(
  null,
  (_, set, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    set(
      specificMonthToolReplacementsRW,
      LazySeq.of(log)
        .filter(
          (e) => e.type === LogType.MachineCycle && !e.startofcycle && !!e.tooluse && e.tooluse.length > 0
        )
        .foldLeft(emptyReplacementsAndUse, addReplacementsFromLog)
    );
  }
);
