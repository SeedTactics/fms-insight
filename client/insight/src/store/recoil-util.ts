/* Copyright (c) 2020, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from "react";
import { RecoilState, useRecoilState, useSetRecoilState } from "recoil";
import produce, { Draft, Immutable } from "immer";

export function useRecoilStateDraft<T>(recoilState: RecoilState<T>): [T, (f: (d: Draft<T>) => void) => void] {
  const [st, setState] = useRecoilState(recoilState);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper: (t: Immutable<Draft<T>>) => Immutable<Draft<T>> = produce((d) => void f(d));
      // convert Immutable<Draft<T>> to T
      setState((mapper as unknown) as (t: T) => T);
    },
    [setState]
  );
  return [st, setDraft];
}

export function useSetRecoilStateDraft<T>(recoilState: RecoilState<T>): (f: (d: Draft<T>) => void) => void {
  const setState = useSetRecoilState(recoilState);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper: (t: Immutable<Draft<T>>) => Immutable<Draft<T>> = produce((d) => void f(d));
      // convert Immutable<Draft<T>> to T
      setState((mapper as unknown) as (t: T) => T);
    },
    [setState]
  );
  return setDraft;
}
