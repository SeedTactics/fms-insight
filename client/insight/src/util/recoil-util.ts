/* Copyright (c) 2021, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from "react";
import {
  RecoilState,
  Snapshot,
  TransactionInterface_UNSTABLE,
  useRecoilState,
  useRecoilTransaction_UNSTABLE,
  useSetRecoilState,
} from "recoil";
import produce, { Draft } from "immer";

export function useRecoilStateDraft<T>(recoilState: RecoilState<T>): [T, (f: (d: Draft<T>) => void) => void] {
  const [st, setState] = useRecoilState(recoilState);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper = produce((d) => void f(d));
      setState(mapper);
    },
    [setState]
  );
  return [st, setDraft];
}

export function useSetRecoilStateDraft<T>(recoilState: RecoilState<T>): (f: (d: Draft<T>) => void) => void {
  const setState = useSetRecoilState(recoilState);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper = produce((d) => void f(d));
      setState(mapper);
    },
    [setState]
  );
  return setDraft;
}

export interface RecoilConduit<T> {
  readonly transform: (trans: TransactionInterface_UNSTABLE, val: T) => void;
}

export function useRecoilConduit<T>({ transform }: RecoilConduit<T>): (val: T) => void {
  return useRecoilTransaction_UNSTABLE((trans) => (val: T) => transform(trans, val));
}

export function conduit<T>(transform: (trans: TransactionInterface_UNSTABLE, val: T) => void): RecoilConduit<T> {
  return { transform };
}

export function applyConduitToSnapshot<T>(snapshot: Snapshot, conduit: RecoilConduit<T>, val: T): Snapshot {
  return snapshot.map((ms) =>
    conduit.transform(
      {
        get: (s) => ms.getLoadable(s).valueOrThrow(),
        set: ms.set,
        reset: ms.reset,
      },
      val
    )
  );
}

export function useImmer<T>(initial: T): [T, (f: (d: Draft<T>) => void) => void] {
  const [state, setState] = React.useState(initial);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper = produce((d) => void f(d));
      setState(mapper);
    },
    [setState]
  );
  return [state, setDraft];
}
