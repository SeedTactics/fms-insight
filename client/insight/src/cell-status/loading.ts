/* Copyright (c) 2021, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { atom, RecoilValueReadOnly, TransactionInterface_UNSTABLE } from "recoil";
import { ICurrentStatus, IHistoricData, ILogEntry, IServerEvent } from "../network/api.js";
import { conduit } from "../util/recoil-util.js";
import { LazySeq } from "@seedtactics/immutable-collections";

import * as simProd from "./sim-production.js";
import * as simUse from "./sim-station-use.js";
import * as schJobs from "./scheduled-jobs.js";
import * as buffers from "./buffers.js";
import * as currentSt from "./current-status.js";
import * as insp from "./inspections.js";
import * as mats from "./material-summary.js";
import * as names from "./names.js";
import * as estimated from "./estimated-cycle-times.js";
import * as tool from "./tool-usage.js";
import * as palCycles from "./pallet-cycles.js";
import * as statCycles from "./station-cycles.js";
import * as toolReplace from "./tool-replacements.js";

export interface ServerEventAndTime {
  readonly evt: Readonly<IServerEvent>;
  readonly now: Date;
  readonly expire: boolean;
}

const lastEventCounterRW = atom<number | null>({
  key: "log-last-event-counter",
  default: null,
});
export const lastEventCounter: RecoilValueReadOnly<number | null> = lastEventCounterRW;

export const onServerEvent = conduit<ServerEventAndTime>(
  (t: TransactionInterface_UNSTABLE, evt: ServerEventAndTime) => {
    simProd.updateLast30JobProduction.transform(t, evt);
    simUse.updateLast30SimStatUse.transform(t, evt);
    schJobs.updateLast30Jobs.transform(t, evt);
    buffers.updateLast30Buffer.transform(t, evt);
    currentSt.updateCurrentStatus.transform(t, evt);
    mats.updateLast30MatSummary.transform(t, evt);
    insp.updateLast30Inspections.transform(t, evt);
    names.updateNames.transform(t, evt);
    tool.updateLast30ToolUse.transform(t, evt);
    palCycles.updateLast30PalletCycles.transform(t, evt);
    toolReplace.updateLastToolReplacements.transform(t, evt);
    statCycles.updateLast30StationCycles.transform(t, evt);

    if (evt.evt.logEntry) {
      const newCntr = evt.evt.logEntry.counter;
      t.set(lastEventCounterRW, (old) => (old === null ? old : Math.max(old, newCntr)));
    }
  }
);

export const onLoadLast30Jobs = conduit<Readonly<IHistoricData>>(
  (t: TransactionInterface_UNSTABLE, historicData: Readonly<IHistoricData>) => {
    simUse.setLast30SimStatUse.transform(t, historicData);
    simProd.setLast30JobProduction.transform(t, historicData);
    schJobs.setLast30Jobs.transform(t, historicData);
    names.setNamesFromLast30Jobs.transform(t, historicData);
  }
);

export const onLoadLast30Log = conduit<ReadonlyArray<Readonly<ILogEntry>>>(
  (t: TransactionInterface_UNSTABLE, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    estimated.setLast30EstimatedCycleTimes.transform(t, log);
    buffers.setLast30Buffer.transform(t, log);
    insp.setLast30Inspections.transform(t, log);
    mats.setLast30MatSummary.transform(t, log);
    names.setNamesFromLast30Evts.transform(t, log);
    tool.setLast30ToolUse.transform(t, log);
    palCycles.setLast30PalletCycles.transform(t, log);
    toolReplace.setLast30ToolReplacements.transform(t, log);
    statCycles.setLast30StationCycles.transform(t, log);

    const newCntr = LazySeq.of(log).maxBy((x) => x.counter)?.counter ?? null;
    t.set(lastEventCounterRW, (oldCntr) => (oldCntr === null ? newCntr : Math.max(oldCntr, newCntr ?? -1)));
  }
);

export const onLoadCurrentSt = conduit<Readonly<ICurrentStatus>>(
  (t: TransactionInterface_UNSTABLE, curSt: Readonly<ICurrentStatus>) => {
    currentSt.setCurrentStatus.transform(t, curSt);
  }
);

export const onLoadSpecificMonthJobs = conduit<Readonly<IHistoricData>>(
  (t: TransactionInterface_UNSTABLE, historicData: Readonly<IHistoricData>) => {
    simUse.setSpecificMonthSimStatUse.transform(t, historicData);
    simProd.setSpecificMonthJobProduction.transform(t, historicData);
    schJobs.updateSpecificMonthJobs.transform(t, historicData);
  }
);

export const onLoadSpecificMonthLog = conduit<ReadonlyArray<Readonly<ILogEntry>>>(
  (t: TransactionInterface_UNSTABLE, log: ReadonlyArray<Readonly<ILogEntry>>) => {
    estimated.setSpecificMonthEstimatedCycleTimes.transform(t, log);
    buffers.setSpecificMonthBuffer.transform(t, log);
    insp.setSpecificMonthInspections.transform(t, log);
    mats.setSpecificMonthMatSummary.transform(t, log);
    palCycles.setSpecificMonthPalletCycles.transform(t, log);
    toolReplace.setSpecificMonthToolReplacements.transform(t, log);
    statCycles.setSpecificMonthStationCycles.transform(t, log);
  }
);
