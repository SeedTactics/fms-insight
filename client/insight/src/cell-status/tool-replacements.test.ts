/* Copyright (c) 2021, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { snapshot_UNSTABLE } from "recoil";
import { toRawJs } from "../../test/to-raw-js.js";
import { applyConduitToSnapshot } from "../util/recoil-util.js";
import { onLoadLast30Log, onServerEvent } from "./loading.js";
import { last30ToolReplacements } from "./tool-replacements.js";
import { it, expect } from "vitest";
import {
  fakeMachineCycle,
  fakeNormalToolUsage,
  fakeToolChangeBeforeCycle,
  fakeToolChangeDuringCycle,
} from "../../test/events.fake.js";
import { addMinutes, addMonths } from "date-fns";
import { LogEntry } from "../network/api.js";

const machCycle = {
  part: "abc",
  proc: 1,
  program: "prog",
  elapsedMin: 1,
  activeMin: 1,
};

it("calculates last 30 tool replacements", () => {
  const start = new Date(Date.UTC(2022, 8, 25, 3, 4, 5));
  const evts = [
    // start with normal usages on machines 4 and 6 with the same tools
    ...fakeMachineCycle({
      ...machCycle,
      counter: 1,
      time: start,
      mcNum: 4,
      tooluse: [
        fakeNormalToolUsage({ tool: "aaa", pocket: 1, minsAtEnd: 40 }),
        fakeNormalToolUsage({ tool: "bbb", pocket: 2, minsAtEnd: 50 }),
        fakeNormalToolUsage({ tool: "ccc", pocket: 3, minsAtEnd: 60 }),
        fakeNormalToolUsage({ tool: "aaa", pocket: 4, minsAtEnd: 70 }),
      ],
    }),
    ...fakeMachineCycle({
      ...machCycle,
      counter: 2,
      time: addMinutes(start, 10),
      mcNum: 6,
      tooluse: [
        fakeNormalToolUsage({ tool: "aaa", pocket: 1, minsAtEnd: 140 }),
        fakeNormalToolUsage({ tool: "bbb", pocket: 2, minsAtEnd: 150 }),
        fakeNormalToolUsage({ tool: "ccc", pocket: 3, minsAtEnd: 160 }),
        fakeNormalToolUsage({ tool: "aaa", pocket: 4, minsAtEnd: 170 }),
      ],
    }),

    // now a couple changes on machine 4
    ...fakeMachineCycle({
      ...machCycle,
      counter: 3,
      time: addMinutes(start, 20),
      mcNum: 4,
      tooluse: [
        fakeToolChangeBeforeCycle({ tool: "aaa", pocket: 1, minsAtEnd: 22 }),
        fakeToolChangeDuringCycle({ tool: "bbb", pocket: 2, use: 16, life: 50, minsAtEnd: 5 }),
        fakeNormalToolUsage({ tool: "ccc", pocket: 3, minsAtEnd: 66 }),
        fakeNormalToolUsage({ tool: "aaa", pocket: 4, minsAtEnd: 77 }),
      ],
    }),
  ];

  const snapshot = applyConduitToSnapshot(snapshot_UNSTABLE(), onLoadLast30Log, evts);
  let tools = snapshot.getLoadable(last30ToolReplacements).valueOrThrow();
  expect(toRawJs(tools)).toMatchSnapshot("tools");

  // now changes on machine 6
  const machEndMc6 = fakeMachineCycle({
    ...machCycle,
    counter: 4,
    time: addMinutes(start, 30),
    mcNum: 6,
    tooluse: [
      fakeToolChangeBeforeCycle({ tool: "aaa", pocket: 1, minsAtEnd: 32 }),
      fakeToolChangeBeforeCycle({ tool: "bbb", pocket: 2, minsAtEnd: 52 }),
      fakeToolChangeDuringCycle({ tool: "ccc", pocket: 3, use: 10, life: 50, minsAtEnd: 7 }),
      fakeNormalToolUsage({ tool: "aaa", pocket: 4, minsAtEnd: 70 }),
    ],
  })[1];

  const afterAdd = applyConduitToSnapshot(snapshot, onServerEvent, {
    now: addMinutes(start, 30),
    evt: { logEntry: new LogEntry(machEndMc6) },
    expire: true,
  });

  tools = afterAdd.getLoadable(last30ToolReplacements).valueOrThrow();
  expect(toRawJs(tools)).toMatchSnapshot("after machine 6 changes");

  // now another change on machine 4, far in the future so it should filter
  const machEndMc4 = fakeMachineCycle({
    ...machCycle,
    counter: 5,
    time: addMonths(start, 2),
    mcNum: 4,
    tooluse: [
      fakeNormalToolUsage({ tool: "aaa", pocket: 1, minsAtEnd: 44 }),
      fakeToolChangeBeforeCycle({ tool: "bbb", pocket: 2, minsAtEnd: 15 }),
      fakeNormalToolUsage({ tool: "ccc", pocket: 3, minsAtEnd: 66 }),
      fakeToolChangeBeforeCycle({ tool: "aaa", pocket: 4, minsAtEnd: 16 }),
    ],
  })[1];

  const afterAdd2 = applyConduitToSnapshot(afterAdd, onServerEvent, {
    now: addMonths(start, 2),
    evt: { logEntry: new LogEntry(machEndMc4) },
    expire: true,
  });
  tools = afterAdd2.getLoadable(last30ToolReplacements).valueOrThrow();
  expect(toRawJs(tools)).toMatchSnapshot("after machine 4 changes");
});
