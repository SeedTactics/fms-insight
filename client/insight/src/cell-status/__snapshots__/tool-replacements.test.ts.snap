// Vitest Snapshot v1

exports[`calculates last 30 tool replacements > after machine 4 changes 1`] = `
Map {
  "MC #4" => Map {
    "5" => {
      "replacements": [
        {
          "cntAtReplacement": null,
          "pocket": 2,
          "tool": "bbb",
          "type": "ReplaceBeforeCycleStart",
          "useAtReplacement": 5,
        },
        {
          "cntAtReplacement": null,
          "pocket": 4,
          "tool": "aaa",
          "type": "ReplaceBeforeCycleStart",
          "useAtReplacement": 77,
        },
      ],
      "time": 2022-10-28T03:04:05.000Z,
    },
  },
}
`;

exports[`calculates last 30 tool replacements > after machine 6 changes 1`] = `
Map {
  "MC #4" => Map {
    "3" => {
      "replacements": [
        {
          "cntAtReplacement": null,
          "pocket": 1,
          "tool": "aaa",
          "type": "ReplaceBeforeCycleStart",
          "useAtReplacement": 40,
        },
        {
          "pocket": 2,
          "tool": "bbb",
          "totalCntAtBeginningOfCycle": null,
          "totalCntAtEndOfCycle": null,
          "totalUseAtBeginningOfCycle": 39,
          "totalUseAtEndOfCycle": 5,
          "type": "ReplaceInCycle",
        },
      ],
      "time": 2022-09-25T03:24:05.000Z,
    },
  },
  "MC #6" => Map {
    "4" => {
      "replacements": [
        {
          "cntAtReplacement": null,
          "pocket": 1,
          "tool": "aaa",
          "type": "ReplaceBeforeCycleStart",
          "useAtReplacement": 140,
        },
        {
          "cntAtReplacement": null,
          "pocket": 2,
          "tool": "bbb",
          "type": "ReplaceBeforeCycleStart",
          "useAtReplacement": 150,
        },
        {
          "pocket": 3,
          "tool": "ccc",
          "totalCntAtBeginningOfCycle": null,
          "totalCntAtEndOfCycle": null,
          "totalUseAtBeginningOfCycle": 47,
          "totalUseAtEndOfCycle": 7,
          "type": "ReplaceInCycle",
        },
      ],
      "time": 2022-09-25T03:34:05.000Z,
    },
  },
}
`;

exports[`calculates last 30 tool replacements > tools 1`] = `
Map {
  "MC #4" => Map {
    "3" => {
      "replacements": [
        {
          "cntAtReplacement": null,
          "pocket": 1,
          "tool": "aaa",
          "type": "ReplaceBeforeCycleStart",
          "useAtReplacement": 40,
        },
        {
          "pocket": 2,
          "tool": "bbb",
          "totalCntAtBeginningOfCycle": null,
          "totalCntAtEndOfCycle": null,
          "totalUseAtBeginningOfCycle": 39,
          "totalUseAtEndOfCycle": 5,
          "type": "ReplaceInCycle",
        },
      ],
      "time": 2022-09-25T03:24:05.000Z,
    },
  },
}
`;
